File: pycore/pyutils/ai_cluster/openai_compat/openai_compat_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""OpenAI-compatible chat client shared by most LLM providers."""

from typing import Any, Dict, List, Optional

from pycore.pyfoundations.third_party.api import get_third_package_requests


class OpenAICompatClient:
    """Minimal OpenAI-compatible REST client (chat + models list)."""

    def __init__(
        self,
        api_key: str,
        base_url: str,
        default_model: str = "",
        provider: str = "",
        extra_headers: Optional[Dict[str, str]] = None,
    ):
        self.api_key = api_key
        self.base_url = (base_url or "").rstrip("/")
        self.default_model = default_model
        self.provider = provider
        self.extra_headers = extra_headers or {}
        self.last_error: Optional[str] = None

    def _headers(self) -> Dict[str, str]:
        h = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        h.update(self.extra_headers)
        return h

    def chat_completion(
        self,
        messages: List[Dict[str, str]],
        model: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: int = 2048,
    ) -> Dict[str, Any]:
        requests = get_third_package_requests()
        use_model = model or self.default_model
        try:
            resp = requests.post(
                f"{self.base_url}/chat/completions",
                headers=self._headers(),
                json={
                    "model": use_model,
                    "messages": messages,
                    "temperature": temperature,
                    "max_tokens": max_tokens,
                },
                timeout=90,
            )
            if resp.status_code == 401:
                self.last_error = "HTTP 401 — invalid API key"
                return self._fail(use_model, self.last_error)
            resp.raise_for_status()
            data = resp.json()
            choices = data.get("choices", [])
            if choices:
                text = (choices[0].get("message") or {}).get("content", "")
                if text:
                    return {
                        "success": True,
                        "text": text,
                        "model": use_model,
                        "provider": self.provider,
                        "error": None,
                    }
            self.last_error = "No text in response"
            return self._fail(use_model, self.last_error)
        except Exception as e:
            self.last_error = str(e)
            return self._fail(use_model, self.last_error)

    def _fail(self, model: str, error: str) -> Dict[str, Any]:
        return {
            "success": False,
            "text": "",
            "model": model,
            "provider": self.provider,
            "error": error,
        }

    def list_models(self) -> List[str]:
        requests = get_third_package_requests()
        self.last_error = None
        try:
            resp = requests.get(
                f"{self.base_url}/models",
                headers={k: v for k, v in self._headers().items() if k != "Content-Type"},
                timeout=20,
            )
            if resp.status_code in (401, 403):
                self.last_error = f"HTTP {resp.status_code} — key invalid or forbidden"
                return []
            resp.raise_for_status()
            data = resp.json().get("data", [])
            return [m.get("id", "") for m in data if isinstance(m, dict) and m.get("id")]
        except Exception as e:
            self.last_error = str(e)
            return []

File: poly_apps/laravel_main/app/Services/AiGateway/OpenAiCompatClient.php
<?php

namespace App\Services\AiGateway;

use Illuminate\Support\Facades\Http;

/**
 * Generic OpenAI-compatible REST client (chat + list-models) built on Laravel's
 * Http facade — no SDK, no cURL handles. Most providers in the registry speak
 * the OpenAI dialect (POST {base}/chat/completions, GET {base}/models with a
 * Bearer token), so a single client covers them; bespoke providers (gemini,
 * anthropic, cloudflare, spark) are handled directly in AiChat / AiProbe.
 */
class OpenAiCompatClient
{
    private string $baseUrl;
    private string $apiKey;
    private array $extraHeaders;

    public function __construct(string $baseUrl, string $apiKey, array $extraHeaders = [])
    {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->apiKey = $apiKey;
        $this->extraHeaders = $extraHeaders;
    }

    /** @return array<string, string> */
    private function headers(): array
    {
        return array_merge([
            'Authorization' => 'Bearer ' . $this->apiKey,
            'Content-Type' => 'application/json',
        ], $this->extraHeaders);
    }

    /**
     * One chat completion. Returns the unified shape:
     *   { success: bool, text: string, error: string|null }
     *
     * @param array<int, array{role:string, content:string}> $messages
     */
    public function chatCompletion(array $messages, string $model, int $timeout = 90, int $maxTokens = 2048): array
    {
        try {
            $response = Http::withHeaders($this->headers())
                ->timeout($timeout)
                ->post($this->baseUrl . '/chat/completions', [
                    'model' => $model,
                    'messages' => $messages,
                    'temperature' => 0.7,
                    'max_tokens' => $maxTokens,
                ]);

            if ($response->status() !== 200) {
                $error = 'HTTP ' . $response->status() . ': ' . mb_substr($response->body(), 0, 300);
                $failure = AiRequestFailure::classify($error);
                return ['success' => false, 'text' => '', 'error' => $error, 'error_code' => $failure['code'], 'provider_reached' => true];
            }

            $data = $response->json();
            $text = $data['choices'][0]['message']['content'] ?? '';
            if (is_array($text)) {
                // Some providers return content as an array of parts.
                $text = implode('', array_map(static fn ($p) => is_array($p) ? ($p['text'] ?? '') : (string) $p, $text));
            }
            $text = (string) $text;

            if ($text === '') {
                return ['success' => false, 'text' => '', 'error' => 'Empty response from provider', 'error_code' => 'empty_response', 'provider_reached' => true];
            }
            return ['success' => true, 'text' => $text, 'error' => null, 'error_code' => null, 'provider_reached' => true];
        } catch (\Throwable $e) {
            $failure = AiRequestFailure::classify($e->getMessage());
            return [
                'success' => false,
                'text' => '',
                'error' => $e->getMessage(),
                'error_code' => $failure['code'],
                'provider_reached' => $failure['provider_reached'],
            ];
        }
    }

    /**
     * List model ids. Returns [models[], error|null].
     *
     * @return array{0: string[], 1: string|null}
     */
    public function listModels(?string $modelsUrl = null, int $timeout = 20): array
    {
        $url = $modelsUrl ?: ($this->baseUrl . '/models');
        try {
            $headers = $this->headers();
            unset($headers['Content-Type']);
            $response = Http::withHeaders($headers)->timeout($timeout)->get($url);

            if (in_array($response->status(), [401, 403], true)) {
                return [[], 'HTTP ' . $response->status() . ' — key invalid or forbidden'];
            }
            if ($response->status() !== 200) {
                return [[], 'HTTP ' . $response->status()];
            }

            $body = $response->json();
            // OpenAI shape: { data: [ {id}, ... ] }; GitHub catalog: [ {id}, ... ].
            $rows = $body['data'] ?? (is_array($body) ? $body : []);
            $ids = [];
            foreach ((array) $rows as $row) {
                if (is_array($row)) {
                    $id = $row['id'] ?? ($row['name'] ?? '');
                    if ($id !== '') {
                        $ids[] = (string) $id;
                    }
                }
            }
            return [$ids, null];
        } catch (\Throwable $e) {
            return [[], $e->getMessage()];
        }
    }
}
